File: solutions/cpp/insertion-deletion-BST.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

using std::cout;
using std::endl;
using std::make_unique;
using std::unique_ptr;

// @include
class BinarySearchTree {
 public:
  // @exclude
  bool Empty() const { return !root_.get(); }

  // @include
  bool Insert(int key) {
    if (root_ == nullptr) {
      root_ = make_unique<TreeNode>(TreeNode{key, nullptr, nullptr});
    } else {
      TreeNode *curr = root_.get(), *parent = nullptr;
      while (curr) {
        parent = curr;
        if (key == curr->data) {
          return false;  // key already present, no duplicates to be added.
        } else if (key < curr->data) {
          curr = curr->left.get();
        } else {  // key > curr->data.
          curr = curr->right.get();
        }
      }

      // Inserts key according to key and parent.
      if (key < parent->data) {
        parent->left.reset(new TreeNode{key});
      } else {
        parent->right.reset(new TreeNode{key});
      }
    }
    return true;
  }

  bool Delete(int key) {
    // Find the node with key.
    TreeNode *curr = root_.get(), *parent = nullptr;
    while (curr && curr->data != key) {
      parent = curr;
      curr = key < curr->data ? curr->left.get() : curr->right.get();
    }

    if (!curr) {
      // There's no node with key in this tree.
      return false;
    }

    TreeNode *key_node = curr;
    if (key_node->right) {
      // Finds the minimum of the right subtree.
      TreeNode *r_key_node = key_node->right.get(), *r_parent = key_node;
      while (r_key_node->left) {
        r_parent = r_key_node;
        r_key_node = r_key_node->left.get();
      }
      key_node->data = r_key_node->data;
      // Moves links to erase the node.
      if (r_parent->left.get() == r_key_node) {
        r_parent->left.reset(r_key_node->right.release());
      } else {  // r_parent->right.get() == r_key_node.
        r_parent->right.reset(r_key_node->right.release());
      }
    } else {
      // Updates root_ link if needed.
      if (root_.get() == key_node) {
        root_.reset(key_node->left.release());
      } else {
        if (parent->left.get() == key_node) {
          parent->left.reset(key_node->left.release());
        } else {  // parent->right.get() == key_node.
          parent->right.reset(key_node->left.release());
        }
      }
    }
    return true;
  }
  // @exclude
  int GetRootVal() const { return root_->data; }
  // @include

 private:
  struct TreeNode {
    int data;
    unique_ptr<TreeNode> left, right;
  };

  unique_ptr<TreeNode> root_ = nullptr;
};
// @exclude

int main(int argc, char *argv[]) {
  BinarySearchTree BST;
  assert(BST.Empty() == true);
  assert(BST.Insert(7) == true);
  assert(BST.Insert(8) == true);
  assert(BST.Insert(9) == true);
  assert(BST.Insert(4) == true);
  assert(BST.Insert(3) == true);
  assert(BST.Empty() == false);
  assert(BST.Insert(2) == true);
  assert(BST.Insert(5) == true);
  assert(BST.Delete(7) == true);
  assert(BST.Delete(9) == true);
  // should output 8
  assert(BST.GetRootVal() == 8);
  cout << BST.GetRootVal() << endl;
  assert(BST.Delete(4) == true);
  // should output 8
  assert(BST.GetRootVal() == 8);
  cout << BST.GetRootVal() << endl;
  assert(BST.Delete(8) == true);
  // should output 5
  assert(BST.GetRootVal() == 5);
  cout << BST.GetRootVal() << endl;
  assert(BST.Delete(5) == true);
  assert(BST.Delete(3) == true);
  // should output 2
  assert(BST.GetRootVal() == 2);
  cout << BST.GetRootVal() << endl;
  assert(BST.Delete(2) == true);
  assert(BST.Delete(1) == false);
  assert(BST.Empty() == true);
  BST.Insert(7);
  assert(BST.GetRootVal() == 7);
  BST.Insert(9);
  assert(BST.GetRootVal() == 7);
  BST.Delete(7);
  assert(BST.GetRootVal() == 9);
  return 0;
}

File: solutions/python/insertion_deletion_bst.py
from binary_tree_prototype import BinaryTreeNode


# @include
class BinarySearchTree:

    def __init__(self):
        self._root = None

    # @exclude
    def empty(self):
        return self._root is None

    # @include
    def insert(self, key):
        if self.empty():
            self._root = BinaryTreeNode(key)
        else:
            parent = None
            curr = self._root
            while curr:
                parent = curr
                if key == curr.data:
                    # key already present, no duplicates to be added.
                    return False
                elif key < curr.data:
                    curr = curr.left
                else:  # key > curr.data.
                    curr = curr.right

            # Inserts key according to key and parent.
            if key < parent.data:
                parent.left = BinaryTreeNode(key)
            else:
                parent.right = BinaryTreeNode(key)
        return True

    def delete(self, key):
        # Find the node with key.
        curr = self._root
        parent = None
        while curr and curr.data != key:
            parent = curr
            curr = curr.left if key < curr.data else curr.right

        if not curr:
            # There's no node with key in this tree.
            return False

        key_node = curr
        if key_node.right:
            # Finds the minimum of the right subtree.
            r_key_node = key_node.right
            r_parent = key_node
            while r_key_node.left:
                r_parent = r_key_node
                r_key_node = r_key_node.left
            key_node.data = r_key_node.data
            # Moves links to erase the node.
            if r_parent.left == r_key_node:
                r_parent.left = r_key_node.right
            else:  # r_parent.right == r_key_node.
                r_parent.right = r_key_node.right
        else:
            # Updates _root link if needed.
            if self._root == key_node:
                self._root = key_node.left
            else:
                if parent.left == key_node:
                    parent.left == key_node.left
                else:  # parent.right == key_node.
                    parent.right = key_node.left
        return True
    # @exclude

    def get_root_val(self):
        return self._root.data


def main():
    bst = BinarySearchTree()
    assert bst.empty() is True
    assert bst.insert(7) is True
    assert bst.insert(8) is True
    assert bst.insert(9) is True
    assert bst.insert(4) is True
    assert bst.insert(3) is True
    assert bst.empty() is False
    assert bst.insert(2) is True
    assert bst.insert(5) is True
    assert bst.delete(7) is True
    assert bst.delete(9) is True
    # should output 8
    assert bst.get_root_val() == 8
    print(bst.get_root_val())
    assert bst.delete(4) is True
    # should output 8
    assert bst.get_root_val() == 8
    print(bst.get_root_val())
    assert bst.delete(8) is True
    # should output 5
    assert bst.get_root_val() == 5
    print(bst.get_root_val())
    assert bst.delete(5) is True
    assert bst.delete(3) is True
    # should output 2
    assert bst.get_root_val() == 2
    print(bst.get_root_val())
    assert bst.delete(2) is True
    assert bst.delete(1) is False
    assert bst.empty() is True
    bst.insert(7)
    assert bst.get_root_val() == 7
    bst.insert(9)
    assert bst.get_root_val() == 7
    bst.delete(7)
    assert bst.get_root_val() == 9


if __name__ == '__main__':
    main()
